flang/lib/Lower/OpenMP.cpp

```diff
@@ -88,29 +116,64 @@ genOMP(Fortran::lower::AbstractConverter &absConv,
+        ifClauseOperand = fir::getBase(
+            absConv.genExprValue(*Fortran::semantics::GetExpr(expr)));
+      } else if (const auto &numThreadsClause =
+                     std::get_if<Fortran::parser::OmpClause::NumThreads>(
+                         &clause.u)) {
         // OMPIRBuilder expects `NUM_THREAD` clause as a `Value`.
-        numThreads = fir::getBase(absConv.genExprValue(
+        numThreadsClauseOperand = fir::getBase(absConv.genExprValue(
             *Fortran::semantics::GetExpr(numThreadsClause->v)));
+      } else if (const auto &defaultClause =
+                     std::get_if<Fortran::parser::OmpDefaultClause>(
+                         &clause.u)) {
+        TODO();
+      } else if (const auto &privateClause =
+                     std::get_if<Fortran::parser::OmpClause::Private>(
+                         &clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList = privateClause->v;
+        genObjectList(ompObjectList, absConv, privateClauseOperands);
+      } else if (const auto &firstprivateClause =
+                     std::get_if<Fortran::parser::OmpClause::Firstprivate>(
+                         &clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList =
+            firstprivateClause->v;
+        genObjectList(ompObjectList, absConv, firstprivateClauseOperands);
+      } else if (const auto &sharedClause =
+                     std::get_if<Fortran::parser::OmpClause::Shared>(
+                         &clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList = sharedClause->v;
+        genObjectList(ompObjectList, absConv, sharedClauseOperands);
+      } else if (const auto &copyinClause =
+                     std::get_if<Fortran::parser::OmpClause::Copyin>(
+                         &clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList = copyinClause->v;
+        genObjectList(ompObjectList, absConv, copyinClauseOperands);
+      } else if (const auto &procBindClause =
+                     std::get_if<Fortran::parser::OmpProcBindClause>(
+                         &clause.u)) {
+        TODO();
       }
     }
     llvm::ArrayRef<mlir::Type> argTy;
-    Attribute defaultValue, procBindValue;
     // Create and insert the operation.
-    // Create the Op with empty ranges for clauses that are yet to be lowered.
     auto parallelOp = firOpBuilder.create<mlir::omp::ParallelOp>(
-        currentLocation, argTy, Value(), numThreads,
-        defaultValue.dyn_cast_or_null<StringAttr>(), ValueRange(), ValueRange(),
-        ValueRange(), ValueRange(),
-        procBindValue.dyn_cast_or_null<StringAttr>());
+        currentLocation, argTy, ifClauseOperand, numThreadsClauseOperand,
+        defaultClauseOperand.dyn_cast_or_null<StringAttr>(),
+        privateClauseOperands, firstprivateClauseOperands, sharedClauseOperands,
+        copyinClauseOperands,
+        procBindClauseOperand.dyn_cast_or_null<StringAttr>());
```

Comment:
You can move the handling of proc_bind and default clause after the op create so you can directly set attribute on it. The code below should get you there. Didn't build it so might need some changes. Also I guess there should be helper functions generated by MLIR TableGen to get the name of the Attr and not hardcoding it 

```suggestion
        procBindClauseOperand.dyn_cast_or_null<StringAttr>());
        // Clauses mapped to attributes. 
        for (const auto &clause : parallelOpClauseList.v) {
          if (const auto &procBindClause =
                      std::get_if<Fortran::parser::OmpProcBindClause>(
                          &clause.u)) {
           switch (procBindClause.v) {
           case OmpProcBindClause::Type::Close:
             parallelOp.setAttr("proc_bind_val", llvm::omp::OMP_PROC_BIND_close); 
             break;
           case OmpProcBindClause::Type::Master:
             parallelOp.setAttr("proc_bind_val", llvm::omp::OMP_PROC_BIND_master);
             break;
           case OmpProcBindClause::Type::Spread:
             parallelOp.setAttr("proc_bind_val", llvm::omp::OMP_PROC_BIND_spread);
             break;
           } else if (const auto &defaultClause =
                      std::get_if<Fortran::parser::OmpDefaultClause>(
                          &clause.u)) {
             switch (defaultClause.v) {
             case OmpDefaultClause::Type::Private:
               parallelOp.setAttr("default_val", mlir::omp::ClauseDefault::ClauseDefaultPrivate);
               break;
             case OmpDefaultClause::Type::Firstprivate:
               parallelOp.setAttr("default_val", mlir::omp::ClauseDefault::ClauseDefaultFirstprivate);
               break;
             case OmpDefaultClause::Type::Shared:
               parallelOp.setAttr("default_val", mlir::omp::ClauseDefault::ClauseDefaultShared);
               break;
             case OmpDefaultClause::Type::None:
               parallelOp.setAttr("default_val", mlir::omp::ClauseDefault::ClauseDefaultNone);
               break;
             }
           }
         }
       }
```

Reply:
Thanks for the suggestion!, I also looked for other solutions.
This one doesn't work since `mlir::omp::ClauseDefault::ClauseDefaultNone` doesn't implicitly converts to `Attribute`.
```
parallelOp.setAttr("default_val", mlir::omp::ClauseDefault::ClauseDefaultNone);
```
Maybe these 2 attributes based clauses deserves a fresh look separately.

Reply:
As I mentioned, there is probably helper functions to translate enum to string for that purpose. This should be in the tablegen generated file for the OpenMP dialect.